runtime/lua/vim/lsp/client.lua

```diff
@@ -227,51 +271,105 @@ local function process_client_config(config)
+local function get_trace(trace)
+  local valid_traces = {
+    off = 'off',
+    messages = 'messages',
+    verbose = 'verbose',
+  }
+  return trace and valid_traces[trace] or 'off'
+end
 
-  if not config.name and type(config.cmd) == 'table' then
-    config.name = config.cmd[1] and vim.fs.basename(config.cmd[1]) or nil
+--- @generic T
+--- @param t T
+--- @return T
+local function copy(t)
+  return vim.deepcopy(t, false)
+end
+
+--- @param id integer
+--- @param config lsp.ClientConfig
+--- @return string
+local function get_name(id, config)
+  local name = config.name
+  if name then
+    return name
+  end
+
+  if type(config.cmd) == 'table' and config.cmd[1] then
+    return assert(vim.fs.basename(config.cmd[1]))
   end
 
-  config.offset_encoding = validate_encoding(config.offset_encoding)
-  config.flags = config.flags or {}
-  config.settings = config.settings or {}
-  config.handlers = config.handlers or {}
+  return tostring(id)
+end
 
-  -- By default, get_language_id just returns the exact filetype it is passed.
-  --    It is possible to pass in something that will calculate a different filetype,
-  --    to be sent by the client.
-  config.get_language_id = config.get_language_id or function(_, filetype)
-    return filetype
+--- @param workspace_folders lsp.WorkspaceFolder[]?
+--- @param root_dir string?
+--- @return lsp.WorkspaceFolder[]?
+local function get_workspace_folders(workspace_folders, root_dir)
+  if workspace_folders then
+    return copy(workspace_folders)
+  end
+  if root_dir then
+    return {
+      {
+        uri = vim.uri_from_fname(root_dir),
+        name = string.format('%s', root_dir),
+      },
+    }
   end
+end
 
-  config.capabilities = config.capabilities or lsp.protocol.make_client_capabilities()
-  config.commands = config.commands or {}
+--- @generic T
+--- @param x elem_or_list<T>?
+--- @return T[]
+local function ensure_list(x)
+  if type(x) == 'table' then
+    return x
+  end
+  return { x }
 end
 
 --- @package
 --- @param config lsp.ClientConfig
 --- @return lsp.Client?
-function Client.start(config)
-  process_client_config(config)
+function Client.create(config)
+  validate_config(config)
 
   client_index = client_index + 1
   local id = client_index
-
-  local name = config.name or tostring(id)
+  local name = get_name(id, config)
 
   --- @class lsp.Client
   local self = {
     id = id,
     config = config,
-    handlers = config.handlers,
-    offset_encoding = config.offset_encoding,
+    handlers = copy(config.handlers or {}),
+    offset_encoding = validate_encoding(config.offset_encoding),
     name = name,
     _log_prefix = string.format('LSP[%s]', name),
     requests = {},
     attached_buffers = {},
     server_capabilities = {},
-    dynamic_capabilities = vim.lsp._dynamic.new(id),
-    commands = config.commands, -- Remove in Nvim 0.11
+    dynamic_capabilities = lsp._dynamic.new(id),
+    commands = copy(config.commands or {}),
+    settings = copy(config.settings or {}),
```

Comment:
I think settings might need to be a reference too, otherwise I suspect it could break configuration plugins.
Currently they can mutate the settings and call `client.notify('workspace/didChangeConfiguration')` to notify the server of a change. That depends on the `workspace_configuration` handler returning the same settings.

Reply:
Hmm. I guess the client doesn't need to edit this table so making it a reference should be fine. I'll do the same for the other fields and remove `copy()`.